contracts/lib/OnChainIdentity.sol

```diff
@@ -0,0 +1,275 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.16;
+pragma abicoder v2;
+
+import "../interfaces/IState.sol";
+import "../lib/SmtLib.sol";
+import "../lib/Poseidon.sol";
+import "../lib/GenesisUtils.sol";
+
+// /**
+//  * @dev Contract managing onchain identity
+//  */
+library OnChainIdentity {
+    using SmtLib for SmtLib.Data;
+
+    /**
+     * @dev Identity data
+     * Issuer Id
+     * Identity identifier
+     * Identity state
+     * State contract
+     * Mapping of roots by state
+     */
+    struct IdentityData {
+        uint256 identityId;
+        uint256 identityLatestState;
+        bool isOldStateGenesis;
+        IState stateContract;
+        mapping(uint256 => Roots) rootsByState;
+        // This empty reserved space is put in place to allow future versions
+        // of the SMT library to add new Data struct fields without shifting down
+        // storage of upgradable contracts that use this struct as a state variable
+        // (see https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#storage-gaps)
+        uint256[45] __gap;
+    }
+
+    /**
+     * @dev SMT addresses
+     */
+    struct Trees {
+        SmtLib.Data claimsTree;
+        SmtLib.Data revocationsTree;
+        SmtLib.Data rootsTree;
+    }
+
+    /**
+     * @dev roots used in last State Transition
+     */
+    struct LastTreeRoots {
+        uint256 claimsRoot;
```

Comment:
You don't need additional `LastTreeRoots` struct that is equal to `Roots` struct. Just use `Roots` struct.